### DISCUSSION TITLE
Preferred convention to store expressions

### DISCUSSION THREAD

**Discussion by tinusd** [casadi/casadi#3303] [Q&A] [answered]
Hello there

Suppose I have the following function:
$f(a,b) = (\cos(a-b))^2$.

My input is $x \in \mathbb{R}^{n}$ and my output is $y \in \mathbb{R}^{n-1}$.
The $i$-th output is $y(i) = f(x(i+1),x(i))$

What would be the best convention to store the $n-1$ expressions of $f$ into $y$? The gradients with respect to $x$ will later be computed. Also, code generation will later be used. 

I have thought of two ways to do this:
**Option 1: cells**
```
import casadi.*;

%% define problem input and output
n = 4;
x = MX.sym('x',n,1);
y = cell(n-1,1);

%% subfunction
a = MX.sym('a');
b = MX.sym('b');
f = Function('f',{a,b},{cos(a-b)^2});

%% compute y
for i=1:n-1  
    y{i} = f(x(i+1),x(i));  
end

y_func = Function('y_func',{x},{[y{:}]'});
```

**Option 2: MX dense vector**
```
import casadi.*;

%% define problem input and output
n = 4;
x = MX.sym('x',n,1);
y = MX.zeros(n-1,1);

%% subfunction
a = MX.sym('a');
b = MX.sym('b');
f = Function('f',{a,b},{cos(a-b)^2});

%% compute y
for i=1:n-1  
    y(i) = f(x(i+1),x(i));  
end

y_func = Function('y_func',{x},{y});
```

Would any of these two options be preferred above the other?

Thanks a lot in advance!

**Comment by jgillis** [comment, marked as answer]
The cell approach leads to a slightly smaller expression graph, and we promote that style in our examples.
Option two can easily go wrong: if you start out with a sparse vector by accident (`y=MX(n-1,1)`) the runtime becomes `O(n^2)`.

Note that your code is also suitable for `map` ( https://web.casadi.org/docs/#for-loop-equivalents ) for further reduction in expression graph size (and hence codegen size)

**Reply by tinusd** [reply]
Thanks a lot, Joris!

Indeed, I am actually already using the map-function to replace the for-loops in my code :) It helped a lot when it came to codegen, speeding it up by an order of magnitude. (The above example is just something I cobbled together.)
